cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -446,7 +503,23 @@ private X509Certificate cert(Secret secret, String key)  {
         }
 
         Future<ReconciliationState> zkRollingUpdate() {
-            return withVoid(zkSetOperations.maybeRollingUpdate(zkDiffs.resource(), zkForcedRestart));
+            if (log.isDebugEnabled()) {
+                String reason = "";
+                if (clusterCaCertRenewed) {
+                    reason += "cluster CA certificate renewal, ";
+                }
+                if (clusterCaCertsRemoved) {
+                    reason += "cluster CA certificate removal, ";
+                }
+                if (zkAncillaryCmChange) {
+                    reason += "ancillary CM change, ";
```

Comment:
What if it is because of the change in the StatefulSet template? That seems to be missing among the reasons.

Reply:
That's currently handled internally within the `StatefulSetOperations`, the `KafkaAssemblyOperator` doesn't itself know. And that's why the method is called `StatefulSetOperations.maybeRollingUpdate()`. It should be possible to refactor that (it's really ugly having the decision spread between the two classes), but not in this PR.